<a href="https://colab.research.google.com/github/culer5/workspaces-student-support/blob/master/Our_First_ML_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Our First ML Model!

Welcome to Google Colab! Today, we're going to build a simple "prediction machine," better known as a machine learning model. You're looking at a Google Colab Notebook, an online environment for running code that works just like a Google Doc!

## Your goal

Simple - work through the notebook.  You don't need to understand the code, add anything to the code, do any math, or anything other than press some buttons.  We've got you all set!

Your goal is to get a small sense of what basic machine learning looks like.

## The Python goal

Our goal is to see if we can predict one thing based on another. In this example, we explore if we can we predict a student's test score based on how many hours the student studied.

## Okay, what else should I know?

The fancy name for this is model is Linear Regression. But all it really means is 'drawing the best possible straight line through a bunch of data points.'  That's it!

We'll use a tiny, simple dataset: 'Hours Studied' vs. 'Test Score.'

Along the way, we'll provide an explanation of what is happening, but again, you won't necessarily need to know any of it if you don't want to.

# Step 1: Prepping Our Notebook
This is a "text cell" It's where you can write plain text.
Below is a "code cell." We'll write our Python code there.

Read the explanation on lines starting with `#`\
The actual code is written on lines without the `#`

If you don't have experience with this, don't edit anything! Just run the cell as is using the arrow button in the upper-left of the code cell

In [1]:
# 1. We import 'pandas' to work with data in tables (like a spreadsheet).
#    We give it a nickname 'pd' to type less.
import pandas as pd

# 2. We import 'plotly.express' to make pretty, interactive graphs.
#    We give it the nickname 'px'.
import plotly.express as px

# 3. This is the "brain" of our model! We import the LinearRegression
#    "recipe" from a famous AI toolkit called 'scikit-learn'.
from sklearn.linear_model import LinearRegression

# 4. Now, all you have to do is hover over this text and click the arrow to the left of line #1
#    It will take a second, but all the code in this cell will run.

# Step 2: Data Creation
Instead of uploading a dataset (which can cause errors), just create the data directly in the Colab.

In [2]:
# Let's make a simple table (called a "DataFrame") with our data.
# We have two columns: 'Hours_Studied' and 'Test_Score'.

data = {
    'Hours_Studied': [1, 2, 3, 4, 5, 7, 8, 10],
    'Test_Score':    [55, 62, 68, 70, 78, 85, 90, 95]
}

df = pd.DataFrame(data)

# Let's look at our data!
print("Our data table:")
print(df)

Our data table:
   Hours_Studied  Test_Score
0              1          55
1              2          62
2              3          68
3              4          70
4              5          78
5              7          85
6              8          90
7             10          95


# Step 3: Look at the Data (The Most Important Step!)
Always visualize before modeling! This helps you understand what your data look like.

In [3]:
# A picture is worth a thousand words. Let's plot our data.
# We want 'Hours_Studied' on the x-axis (bottom)
# and 'Test_Score' on the y-axis (side).

fig = px.scatter(df,
                 x='Hours_Studied',
                 y='Test_Score',
                 title='Does More Studying Lead to Higher Scores?')
fig.show()

See the graph above? (You can hover over the dots!)

You can clearly see a trend: the more hours someone studies, the higher their test score tends to be.

Our ML's job is to find the single best line that cuts through these dots. This line will be our 'predictor.'

#Step 4: Prepare Data for the "Brain"

In [4]:
# Our ML model needs to know two things:
# 1. The "Clue" (or "Input"). We call this 'X'.
# 2. The "Answer" (or "Output"). We call this 'y'.

# The "Clue" is the 'Hours_Studied'.
# (We use double brackets [[]] because the model likes it in a "table" format).
X = df[['Hours_Studied']]

# The "Answer" is the 'Test_Score'.
y = df['Test_Score']

print("This is our 'Clue' (X):")
print(X)
print("\nThis is our 'Answer' (y):")
print(y)

This is our 'Clue' (X):
   Hours_Studied
0              1
1              2
2              3
3              4
4              5
5              7
6              8
7             10

This is our 'Answer' (y):
0    55
1    62
2    68
3    70
4    78
5    85
6    90
7    95
Name: Test_Score, dtype: int64


#Step 5: Create & Train the Model
This is the "magic" part. Look how simple it is!

In [5]:
# 1. Create a new, blank "brain" (model) from our recipe.
model = LinearRegression()

# 2. "Train" the brain.
# This is the most important line! We're telling the model
# to "fit" itself to our X and y data. It's looking at all
# the dots and figuring out the best line.
model.fit(X, y)

print("Done! The model has been trained.")

Done! The model has been trained.


#Step 6: What Did the Model "Learn?"
So, what did the model learn? It learned the formula for a line: `y = mx + b`

In our case, that's: Score = (Slope * Hours) + Intercept

The Intercept (b) is the "starting point." What score would you get if you studied for 0 hours?

The Slope (m) is the "steepness." For every 1 extra hour you study, how many extra points do you get?

In [6]:
# Let's see what the model learned!

# This is the 'b' (the Intercept)
intercept = model.intercept_

# This is the 'm' (the Slope)
# It's in an "array" (a list), so we grab the first item [0]
slope = model.coef_[0]

print(f"Model's Intercept (b): {intercept:.2f}")
print(f"Model's Slope (m): {slope:.2f}")

print(f"\nThis means the model learned the formula:")
print(f"Score = ({slope:.2f} * Hours_Studied) + {intercept:.2f}")

Model's Intercept (b): 53.10
Model's Slope (m): 4.46

This means the model learned the formula:
Score = (4.46 * Hours_Studied) + 53.10


Intercept (b) ≈ 53.10: The model thinks a student who studies 0 hours will get a score of about 53.10.

Slope (m) ≈ 4.46: The model thinks that for every 1 extra hour of studying, a student's score will go up by 4.46 points.

See? That's it! That's the 'intelligence.' It's not magic, it's just a formula learned from the data.

# Step 7: Let's Make a Prediction!

In [7]:
# Let's predict a score for someone who studied for 6 hours.
# (Note: We have to give it the 6 in double brackets [[6]])

hours_to_predict = pd.DataFrame({'Hours_Studied': [6]})
predicted_score = model.predict(hours_to_predict)

print(f"The model predicts a score of {predicted_score[0]:.2f} for someone who studied 6 hours.")

# You can even check the math!
# (Slope * 6) + Intercept
math_check = (slope * 6) + intercept
print(f"Our manual math check: {math_check:.2f} (It matches!)")

The model predicts a score of 79.83 for someone who studied 6 hours.
Our manual math check: 79.83 (It matches!)


In [8]:
# Let's draw the same scatter plot as before...
fig = px.scatter(df,
                 x='Hours_Studied',
                 y='Test_Score',
                 title="Model's 'Best Fit' Line vs. Real Data")

# ...but this time, let's add the model's predictions (the line).
# We'll ask the model to predict the score for *every* X value in our data.
fig.add_trace(px.line(x=df['Hours_Studied'],
                      y=model.predict(X)).data[0])

fig.show()

Look at that! The dots are the real data. The blue line is our model's 'best guess' line.

It doesn't hit every dot perfectly, and that's normal! People are more complicated than a simple line.

But this line is the best summary of the trend. It's the most helpful 'predictor' we can make from this data.

# Wrapping it up!
Congratulations! You have successfully built, trained, and interpreted your very first machine learning model.

What you learned:


*   ML/AI models learn from data (the dots).

*   "Training" just means finding a pattern (the best-fit line).

*   The result isn't magic, it's a formula (like `Score = (4.34 * Hours) + 50.85)`.

That process you just ran—'fitting a model to data'—is the foundation for almost all modern AI.

The leap from what you just did to what ChatGPT does isn't magic; it's just scale.

You taught a model to find the "best fit line." A larger model, like the one used for ChatGPT, is taught to find the "best fit word." A self-driving car's AI is taught to find the "best fit steering action."

They are all just finding patterns and making predictions based on data. You've just proven the fundamental concept. You now know the "trick" behind the "magic!"